### PR TITLE
Update to_dict_of_dict edge_data

### LINKED ISSUE #4305: Question: MultiGraph with to_dict_of_dicts with edge data
I was reviewing the tests for `nx.convert.to_dict_of_dicts` which led to a question about what the expected behavior of `nx.to_dict_of_dicts(G)` is when G is a MultiGraph and `edge_data` is not None.

Things seem to make sense when `edge_data` is None:

```python
>>> G = nx.MultiGraph()
>>> G.add_edge(0, 1, key='a')
'a'
>>> G.add_edge(0, 1, key='b')
'b'
>>> nx.to_dict_of_dicts(G, edge_data=None)
{0: {1: {'a': {}, 'b': {}}}, 1: {0: {'a': {}, 'b': {}}}}
```

In the above example, if I call `nx.to_dict_of_dicts(G, edge_data={'weight': 1})`, my naive expectation was that i would get

```python
{0: {1: {'a': {'weight': 1}, 'b': {'weight': 1}}}, 1: {0: {'a': {'weight': 1}, 'b': {'weight': 1}}}}
```

Instead, you get:

```python
>>> nx.to_dict_of_dicts(G, edge_data={'weight': 1})
{0: {1: {'weight': 1}}, 1: {0: {'weight': 1}}}
```
which seems to clobber the multiple edges. [The code](https://github.com/networkx/networkx/blob/9aee8ef9739913ebd1e34b1be97cd91a7d6cc520/networkx/convert.py#L240-L277) doesn't have any special cases for MultiGraphs, and I couldn't tell from there and/or docstring whether this behavior is expected.

I expect I'm missing something, so I thought I'd ask! Adding examples to the docstring would also be a nice improvement to show the expected behavior.

### PR DESCRIPTION
**Pull request by rossbar**
Closes #4305 .

Updates the documentation of `to_dict_of_dicts` to better describe and provide examples for the `edge_data` kwarg. Also adds tests for the (previously uncovered) case when the `edge_data` kwarg is not `None`.

### CODE REVIEW

**Comment by rossbar** [comment]
Adding the 2.6 milestone as this will improve test coverage of the convert module.

File: networkx/convert.py [hunk 1]
```diff
@@ -248,12 +248,91 @@ def to_dict_of_dicts(G, nodelist=None, edge_data=None):
...
+    Returns
+    -------
+    dod : dict
+       A nested dictionary representation of `G`. Note that the level of
+       nesting depends on the type of `G` and the value of `edge_data`
+       (see Examples).
+
+    See Also
+    --------
+    from_dict_of_dicts, to_dict_of_lists
```
Comment (Erotemic) [inline networkx/convert.py:268]: Should these have rst style references? e.g. 

```
`func`:from_dict_of_dicts`
```

File: networkx/convert.py [hunk 2]
```diff
@@ -248,12 +248,91 @@ def to_dict_of_dicts(G, nodelist=None, edge_data=None):
...
        Use only nodes specified in nodelist
 
-    edge_data : list, optional
-       If provided,  the value of the dictionary will be
-       set to edge_data for all edges.  This is useful to make
-       an adjacency matrix type representation with 1 as the edge data.
-       If edgedata is None, the edgedata in G is used to fill the values.
-       If G is a multigraph, the edgedata is a dict for each pair (u,v).
+    edge_data : singleton, optional
+       If provided, the value of the dictionary will be set to `edge_data` for
```
Comment (Erotemic) [inline networkx/convert.py:252]: Unlike markdown, I believe rst wants inline code to be surrounded by double backticks: e.g.

```
``edge_data``
```

instead of 

```
`edge_data`
```

Not sure how sphinx handles this though. It might be fine.

File: networkx/convert.py [hunk 3]
```diff
@@ -248,12 +248,91 @@ def to_dict_of_dicts(G, nodelist=None, edge_data=None):
...
+    >>> G = nx.Graph()
+    >>> G.add_edges_from(
+    ...     [
+    ...         (0, 1, {'weight': 1.0}),
+    ...         (1, 2, {'weight': 2.0}),
+    ...         (2, 0, {'weight': 1.0}),
+    ...     ]
+    ... )
+    >>> d = nx.to_dict_of_dicts(G)
+    >>> d  # doctest: +SKIP
```
Comment (Erotemic) [inline networkx/convert.py:301]: Looks like another potential use-case for xdoctest 😉

File: networkx/convert.py (hunk 1 above)
Comment (rossbar) [reply networkx/convert.py:268]: They don't need them - the sphinx extensions that handles numpydoc-formatted docstring parsing (napoleon in this case) handles the roles automatically.

You can see the rendered documentation via the `circleci: build artifact` tab in the checks below. Here's a link to the page for `to_dict_of_dicts`: https://629-890377-gh.circle-artifacts.com/0/doc/build/html/reference/generated/networkx.convert.to_dict_of_dicts.html

File: networkx/convert.py (hunk 2 above)
Comment (rossbar) [reply networkx/convert.py:252]: According to the [numpydoc docstring standard](https://numpydoc.readthedocs.io/en/latest/format.html#common-rest-concepts), parameters should still be enclosed in single backticks (default role), not double-backticks (inline literal).

File: networkx/convert.py (hunk 3 above)
Comment (rossbar) [reply networkx/convert.py:301]: I haven't had a chance to follow the xdoctest discussion too closely yet, so I guess I'll take the opportunity to ask here: does xdoctest handle outputs that are formatted differently than the traditional REPL? Using this case as an example, if I run this in the traditional Python REPL I get:

```python
>>> d
{0: {1: {'weight': 1.0}, 2: {'weight': 1.0}}, 1: {0: {'weight': 1.0}, 2: {'weight': 2.0}}, 2: {1: {'weight': 2.0}, 0: {'weight': 1.0}}}
```

vs. the IPython terminal:

```python
In [1]: d
{0: {1: {'weight': 1.0}, 2: {'weight': 1.0}},
 1: {0: {'weight': 1.0}, 2: {'weight': 2.0}},
 2: {1: {'weight': 2.0}, 0: {'weight': 1.0}}}
```

The latter is obviously easier to read, but will unfortunately fail the doctests because it doesn't exactly match REPL formatting. Does `xdoctest` support IPython-formatted outputs? If so, that would be a compelling argument in favor of xdoctest since reprs of nested containers show up pretty frequently in the nx docs

Comment (Erotemic) [reply networkx/convert.py:301]: Woops, I lied. xdoctest doesn't actually handle this case. There is a handler to ignore whitespace, but apparently that doesn't handle newlines. I'll likely fix that in the next version.

I probably would have written this test like this:

```python
    >>> import pprint
    >>> print(pprint.pformat(d))
    {0: {1: {'weight': 1.0}, 2: {'weight': 1.0}},
     1: {0: {'weight': 1.0}, 2: {'weight': 2.0}},
     2: {0: {'weight': 1.0}, 1: {'weight': 2.0}}}
```

but I think that works with regular doctest anyway.

File: networkx/convert.py [hunk 4]
```diff
@@ -248,12 +248,91 @@ def to_dict_of_dicts(G, nodelist=None, edge_data=None):
     nodelist : list
        Use only nodes specified in nodelist
 
-    edge_data : list, optional
-       If provided,  the value of the dictionary will be
-       set to edge_data for all edges.  This is useful to make
-       an adjacency matrix type representation with 1 as the edge data.
-       If edgedata is None, the edgedata in G is used to fill the values.
-       If G is a multigraph, the edgedata is a dict for each pair (u,v).
+    edge_data : singleton, optional
```
Comment (stefanv) [inline networkx/convert.py:251]: singleton -> scalar or similar

File: networkx/convert.py [hunk 5]
```diff
@@ -248,12 +248,91 @@ def to_dict_of_dicts(G, nodelist=None, edge_data=None):
...
+    See Also
+    --------
+    from_dict_of_dicts, to_dict_of_lists
+
+    Notes
+    -----
+    `edge_data` should *not* be a container. For a more custom approach to
+    handling edge data, try::
+
+        dod = {
```
Comment (stefanv) [inline networkx/convert.py:275]: Is this intended to show a custom transformation on `n, nbr, dd`?

It wasn't clear to me what the different short-hands were.

File: networkx/convert.py [hunk 6]
```diff
@@ -248,12 +248,91 @@ def to_dict_of_dicts(G, nodelist=None, edge_data=None):
...
+       nesting depends on the type of `G` and the value of `edge_data`
+       (see Examples).
+
+    See Also
+    --------
+    from_dict_of_dicts, to_dict_of_lists
+
+    Notes
+    -----
+    `edge_data` should *not* be a container. For a more custom approach to
```
Comment (stefanv) [inline networkx/convert.py:272]: Should this warning be provided in the parameter description?

File: networkx/convert.py (hunk 5 above)
Comment (rossbar) [reply networkx/convert.py:275]: Yeah, the idea is to show how one could create a nested-dict structure with custom (potentially non-scalar) edge data. I'll add a sentence at the end to hopefully clarify the intent and expand on the variable names.

**Review by jarrodmillman — APPROVED** [review]
